shiny/ui/_chat.py

```diff
@@ -1325,6 +1376,210 @@ async def _send_custom_message(self, handler: str, obj: ClientMessage | None):
+                    ), "Chat bookmark value must be a list of strings"
+
+                turns_json_str = cast(list[str], value)
+
+                print("restored messages", turns_json_str)
+                print("restored messages", type(turns_json_str))
+                print("restored messages", len(turns_json_str))
+                print("restored messages", turns_json_str[0])
+
+                turns: list[ChatlasTurn[Any]] = [
+                    ChatlasTurn.model_validate_json(turn_json_str)
+                    for turn_json_str in turns_json_str
+                ]
+                client.set_turns(turns)  # pyright: ignore[reportUnknownMemberType]
+
+            get_state = get_chatlas_state
+            set_state = set_chatlas_state
+
+        else:
+            raise ValueError(
+                "Bookmarking requires a client that supports "
+                "`async def get_state(self) -> shiny.types.Jsonifiable` (which returns an object that can be used when bookmarking to save the state of the `client=`) and "
+                "`async def set_state(self, value: Jsonifiable)` (which should restore the `client=`'s state given the `value=`)."
+            )
+
+        resolved_bookmark_id_str = str(self.id)
+        resolved_bookmark_id_msgs_str = resolved_bookmark_id_str + "--msgs"
+
+        session = get_current_session()
+        if session is None or isinstance(session, ExpressStubSession):
+            return BookmarkCancelCallback(lambda: None)
+
+        if session.bookmark.store == "disable":
+            raise ValueError(
+                "Bookmarking requires a `bookmark_store` to be set. Please set `bookmark_store=` in `App()`."
+            )
+
+        # Reset prior bookmarking hooks
+        self._destroy_bookmarks()
+
+        # Must use `root_session` as the id is already resolved. :-/
+        # Using a proxy session would double-encode the proxy-prefix
+        root_session = session.root_scope()
+        root_session.bookmark.exclude.append(self.id + "_user_input")
+
+        if bookmark_on is not None:
+
+            # When ever the bookmark is requested, update the query string (indep of store type)
+            @root_session.bookmark.on_bookmarked
+            async def _(url: str):
+                await session.bookmark.update_query_string(url)
+
+        if bookmark_on == "response":
+
+            @reactive.effect
+            # @reactive.event(lambda: chat.latest_message_stream.result, ignore_init=True)
+            @reactive.event(lambda: self.messages(format=MISSING), ignore_init=True)
+            async def _():
+                messages = self.messages(format=MISSING)
+
+                if len(messages) == 0:
+                    return
+
+                last_message = messages[-1]
+
+                if last_message.get("role") == "assistant":
+                    await session.bookmark()
+
+        @root_session.bookmark.on_bookmark
+        async def cancel_on_bookmark_client(state: BookmarkState):
```

Comment:
Small nit, and feel free to say no, but when I read `cancel` I'm expecting the `def` itself to be about canceling, not the transformed `def`. 

```suggestion
        async def _on_bookmark_client(state: BookmarkState):
```